### PR TITLE
fix 125: add unit tests

### LINKED ISSUE #125: Identify unit testing gaps and fill them
Identify any gaps in unit test coverage and areas that need some more and fill them. Out of the three areas (code generator tool, runtime library, and generated code), the runtime library does need more coverage. The code generator tool probably needs some cleanup and maybe more tests for what remains.

```[tasklist]
### Tasks
- [x] Check that the new `hashcode()` and `equals()` methods are tested in the generated code, including corner cases
- [ ] #28
- [ ] To do - add more specific tasks
```

### PR DESCRIPTION
**Pull request by anthony-swirldslabs**
**Description**:
Adding some unit tests which should improve the coverage for PBJ.

**Related issue(s)**:

Fixes #125

**Notes for reviewer**:
```
pbj-core]$ gr test
...
BUILD SUCCESSFUL in 20s
```

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 1]
```diff
@@ -142,6 +152,16 @@ void testWriteTagSpecialWireType() {
...
+        final int valToWrite = 0;
+        final long positionBefore = bufferedData.position();
+        writeInteger(bufferedData, definition, valToWrite);
+        final long positionAfter = bufferedData.position();
+        assertEquals(positionBefore, positionAfter);
+    }
+
     @Test
     void testWriteInteger_int32() {
         FieldDefinition definition = createFieldDefinition(INT32);
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:167]: This test case writes a random int. If the int happens to be 0 (the default int value), the test would fail. Does it make sense to protect from this as a part of current PR?

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 2]
```diff
@@ -192,6 +212,16 @@ void testWriteInteger_unsupported(FieldType type) {
         assertThrows(RuntimeException.class, () -> writeInteger(bufferedData, definition, RNG.nextInt()));
     }
 
+    @Test
+    void testWriteLong_Zero() {
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:216]: s/Zero/zero/ , or better use something like "default", cause the test case actually verifies how default values are (not) written.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 3]
```diff
@@ -280,19 +310,44 @@ void testWriteBoolean_false() {
...
+
     @Test
     void testWriteEnum() {
         FieldDefinition definition = createFieldDefinition(ENUM);
-        EnumWithProtoMetadata enumWithProtoMetadata = mock(EnumWithProtoMetadata.class);
-        int expectedOrdinal = RNG.nextInt();
-        when(enumWithProtoMetadata.protoOrdinal()).thenReturn(expectedOrdinal);
-        when(enumWithProtoMetadata.protoName()).thenReturn(RANDOM_STRING.nextString());
+        final int expectedOrdinal;
+        EnumWithProtoMetadata enumWithProtoMetadata = mockEnum(expectedOrdinal = RNG.nextInt());
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:324]: I wonder why `expectedOrdinal` is initialized here rather than on the previous line?

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 4]
```diff
@@ -986,6 +1059,268 @@ void testSizeOfDelimited() {
...
+                0xffff,
+                0xfffff,
+                0xffffff,
+                0xfffffff,
+                0xffffffff
+        ));
+        final long finish = bufferedData.position();
+
+        int tag = bufferedData.getVarInt(start, false);
+        assertEquals(WIRE_TYPE_DELIMITED.ordinal(), tag & 0x3);
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1091]: `ProtoParserTools` has a constant for this 0x03 value, it can be used here. Also, the constant seems to be useful not only while parsing, but also on writing, so it can be moved to `ProtoConstants`, maybe?

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 5]
```diff
@@ -986,6 +1059,268 @@ void testSizeOfDelimited() {
...
+        int sizeOfTag = ProtoWriterTools.sizeOfVarInt32(tag);
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1092]: Can `ProtoWriterTools.sizeOfTag()` be used here?

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 6]
```diff
@@ -986,6 +1059,268 @@ void testSizeOfDelimited() {
...
+        assertEquals(expectedSize, size);
+        int sizeOfSize = ProtoWriterTools.sizeOfVarInt32(size);
+
+        assertEquals(finish - start - sizeOfTag - sizeOfSize, size);
+
+        long firstLong = bufferedData.getVarLong(start + sizeOfTag + sizeOfSize, zigZag);
+        assertEquals(0x0f, firstLong);
+    }
+
+    private static interface WriterMethod<T> {
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1154]: I'm not sure, but can `ProtoWriter` interface be reused here?

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java [hunk 7]
```diff
@@ -986,6 +1059,268 @@ void testSizeOfDelimited() {
...
+            assertEquals((definition.number() << 3) | WIRE_TYPE_DELIMITED.ordinal(), tag);
+            int sizeOfTag = ProtoWriterTools.sizeOfVarInt32(tag);
+
+            UnpackedField<T> value = readerMethod.read(bufferedData, start + offset + sizeOfTag);
+            assertEquals(list.get(i), value.value());
+
+            offset += sizeOfTag + value.size();
+        }
+
+        assertEquals(finish,start + offset);
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1322]: Whitespace

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java [hunk 8]
```diff
@@ -11,6 +20,38 @@ protected ReadableSequentialData emptySequence() {
         return new StubbedSequence(new byte[0]);
     }
 
+    @NonNull
+    private ReadableSequentialData throwingSequence() {
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java:24]: This would test `StubbedSequence` only. Instead, it would be great to have this method as abstract in `ReadableSequentialTestBase` and implement it wherever needed. It would test all ReadableSequentialData implementations.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java [hunk 9]
```diff
@@ -11,6 +20,38 @@ protected ReadableSequentialData emptySequence() {
...
+        return new StubbedSequence(new byte[] { 1 }, () -> new EOFException());
+    }
+
+    @Test
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java:28]: These new test cases can be moved to `ReadableSequentialTestBase`, too

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableTestBase.java [hunk 10]
```diff
@@ -1293,12 +1334,30 @@ void readZigZag() {
...
+                    (byte) 0b10000010,
+                    (byte) 0b10000010,
+                    (byte) 0b10000010,
+                    (byte) 0b10000010,
+                    (byte) 0b10000010,
+                    (byte) 0b10000010
+            });
+            // When we try to decode an int, the lack of a "terminator" bit causes a DataEncodingException
+            assertThatThrownBy(() -> seq.readVarInt(false)).isInstanceOf(DataEncodingException.class);
+        }
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableTestBase.java:1355]: Is there a test that calls `readVarInt()` for a sequence of bytes, where more than 5 bytes have continuation bit set, and the last one does not? `readVarLong()` should handle such a sequence just fine, but (to my surprise), `readVarInt()` should not fail either. It should read it as a long and then downcast to int, if I remember correctly.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/buffer/BufferedDataTestBase.java [hunk 11]
```diff
@@ -106,6 +106,22 @@ void sliceThenReadVarInt(final int num) {
         assertEquals(afterSecondIntPos - afterFirstIntPos, slicedBuf.position());
     }
 
+    @ParameterizedTest
+    @ValueSource(ints = {0, 1, 2, 4, 7, 8, 15, 16, 31, 32, 33, 127, 128, 512, 1000, 1024, 4000,
+            16384, 65535, 65536, 65537, 0xFFFFFF, 0x1000000, 0x1000001, 0x7FFFFFFF,
+            -1, -7, -8, -9, -127, -128, -129, -65535, -65536, -0xFFFFFF, -0x1000000, -0x1000001, -0x80000000})
+    @DisplayName("readVar() won't read beyond 10 bytes")
+    void readVarFromLargeBuffer(final int num) {
+        final var buf = allocate(100);
```
Comment (artemananiev) [inline pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/buffer/BufferedDataTestBase.java:115]: Var ints (including longs) may not take more than 10 bytes

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 1 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:167]: Good catch. Addressed it for this and a few more subsequent tests.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 2 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:216]: Changed to `zero`.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 3 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:324]: No particular reason. Moved the assignment to the declaration.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 4 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1091]: That's a very good catch. I think it should actually be 0x7 because there's 3 bits in the wire type. I'll use a common constant for that.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 5 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1092]: Yes, in theory. However, this particular test verifies the bytes literally written to the buffer by a particular method `ProtoWriterTools.writeIntegerList`. So I want to check the raw written bytes only and not rely on other high(er)-level logic implemented in ProtoWriterTools.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 6 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1154]: The ProtoWriter seems to accept an output sink and a single data variable, while this interface would be unable to infer a particular field definition from the given data object alone. So I need this custom interface here that also accepts the field definition.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java (hunk 7 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/ProtoWriterToolsTest.java:1322]: Fixed.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java (hunk 8 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java:24]: Yes, that was my original idea, too. But I found it unfeasible to support this test in one of the descendants of the TestBase - I recall I just couldn't reach the particular branch that would throw the exception from the test that uses this sequence. That's the reason why it's sitting here.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java (hunk 9 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableSequentialDataTest.java:28]: Same reason as above: I didn't find an easy way to reach a branch that would trigger the exception in one of the descendent classes.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableTestBase.java (hunk 10 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/ReadableTestBase.java:1355]: Yes, readVarInt() basically delegates to readVarLong(). So there's no way to meaningfully test this until/unless we actually separate them and start failing the readVarInt on longer sequences. But I guess this would introduce some incompatibilities with Google's implementation (IIRC, right?), that's why we simply delegate for now.

File: pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/buffer/BufferedDataTestBase.java (hunk 11 above)
Comment (anthony-swirldslabs) [reply pbj-core/pbj-runtime/src/test/java/com/hedera/pbj/runtime/io/buffer/BufferedDataTestBase.java:115]: Yes, yes, but as the name of the test suggests it specifically wants to use a longer buffer to check an edge case somewhere in the implementation. 100 is long enough for me.

That edge case may have existed in the old readVarInt implementation and perhaps you removed it when you overhauled its implementation a few weeks ago. However, the test seems good still - it still covers the edge case and/or will cover it if/when we re-introduce it in the future.

**Review by imalygin — APPROVED** [review]
